core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java

```diff
@@ -124,16 +125,54 @@ public String generateLoadBalancerEndpoint(Stack stack) {
         }
     }
 
+    /*
+     *  follows these peferences:
+     *   public fqdn
+     *   public DNS
+     *   public IP
+     *   private FQDN
+     *   private DNS
+     *   private IP
+     */
     public String getLoadBalancerUserFacingFQDN(Long stackId) {
         Set<LoadBalancer> loadBalancers = loadBalancerPersistenceService.findByStackId(stackId).stream()
-            .filter(lb -> StringUtils.isNotEmpty(lb.getFqdn()))
-            .collect(Collectors.toSet());
-        LoadBalancer loadBalancer = loadBalancers.stream()
-            .filter(lb -> LoadBalancerType.PUBLIC.equals(lb.getType()))
-            .findAny()
-            .orElseGet(() -> loadBalancers.stream().findAny().orElse(null));
+                .filter(lb -> StringUtils.isNotEmpty(lb.getDns()) ||
+                        StringUtils.isNotEmpty(lb.getIp()) || StringUtils.isNotEmpty(lb.getFqdn()))
+                .collect(Collectors.toSet());
+        if (loadBalancers.isEmpty()) {
+            return null;
+        }
+        Iterator<LoadBalancer> iterator = loadBalancers.iterator();
+        LoadBalancer best = iterator.next();
+        while (iterator.hasNext()) {
+            LoadBalancer balancer = iterator.next();
+            if (!best.getType().equals(balancer.getType())) {
+                if (best.getType().equals(LoadBalancerType.PRIVATE)) {
+                    best = balancer;
+                }
+                continue;
+            }
+            if (StringUtils.isNotEmpty(balancer.getFqdn()) && StringUtils.isEmpty(best.getFqdn())) {
+                best = balancer;
+            }
 
-        return loadBalancer == null ? null : loadBalancer.getFqdn();
+        }
+        return getBestAddressable(best);
+
+    }
+
+    private String getBestAddressable(LoadBalancer lb) {
+        Preconditions.checkNotNull(lb);
+        if (StringUtils.isNotBlank(lb.getFqdn())) {
+            return lb.getFqdn();
+        }
+        if (StringUtils.isNotBlank(lb.getDns())) {
+            return lb.getDns();
+        }
+        if (StringUtils.isNotBlank(lb.getIp())) {
+            return lb.getIp();
+        }
+        return null;
```

Comment:
how could this happen?

Reply:
I believe all of these could be null, if the loadBalancer had just been initialized before any load balance metadata had been applied to it.

Reply:
@lacikaaa That could theoretically happen if the user is looking in the datalake UI while the stack is still spinning up. When the UI tries to get the user facing FQDN for the CM server, it'll land here. If we land here before the metadata has been collected, we just want to return null as an indication that we don't have the information yet.

Reply:
I was trying to point out a bit indirect way that this code is not reachable. neither this line nor the first if is reachable. if it is please cover with unit tests, as these are not covered yet. If these can't happen why we have these?
```
        if (StringUtils.isNotBlank(lb.getFqdn())) {
            return lb.getFqdn();
        } else if (StringUtils.isNotBlank(lb.getDns())) {
            return lb.getDns();
        } else {
            return lb.getIp();
        }
```
this should be sufficient if you replace the filters from `notEmpty` to `notBlank`

Reply:
The only difference between `isNotBlank` and `isNotEmpty` is `isNotBlank` will return false if the string contains only whitespace, while `isNotEmpty` will return true in that case. They both return false if the string is null or empty.

`testGetLoadBalancerUserFacingFQDN` verifies the first if statement can be reached, `testGetLoadBalancerUserFacingFQDNPublicMissingFQDN` verifies the else/if can be reached, `testGetLoadBalancerUserFacingFQDNIPOnlyMissingFQDN` verifies the final else can be reached, and `testGetLoadBalancerUserFacingFQDNNoFqdnSet` verifies that if none of those are hit then we return null at the end of the method.

Can you explain why you think the code is unreachable?

Reply:
thanks, I know the difference. I just wanted to make sure the filter in `getLoadBalancerUserFacingFQDN` is in sync with the conditions here.
first `if` is unreachable because this will never return `null`:
`loadBalancerPersistenceService.findByStackId(stackId)`

the last null is unreachable because of the filter, which filters out all the LB which doesn't have at least one of these field set.

Reply:
Ah, I understand now. Thanks for clarifying. Since we're returning null in to different places, the unit test isn't accurately verifying which null return statement is being hit, and so isn't verifying what it's meant to. Thanks for that catch!

@yg-gd to that end I think things would be cleaner if you only had a single null return statement. It would also be good to have a unit test case where the LB exists but has null fqdn, dns, and ip, and a separate test case for when there are no load balancers in the stack at all. That's a test case I missed the first go-round, but it'd be good to have.

Reply:
I took @lacikaaa's point In the case where all 3 are null, then it's safe to return the IP without checking since it's the same as returning null. I've left the initial null check in case this is used in the future in some other context where it's possible that lb is null.

Reply:
my opinion still: you should drop that check. it's a private method called from one other. this is unused unnecessary code. if we would code like this, we could have tons of null check in the beginning of all methods. the best is to avoid nulls as soon as possible.

Reply:
A compiler is able to infer that it's not possible for lb to be null at this point and will optimize away this check regardless. What remains is whether or not if it's helpful for humans reading the code to understand that inference.